src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1807,4 +1807,67 @@ else if ("test".equals(configName)) {
                 .isFalse();
     }
 
+    @Test
+    public void testAllExampleMacrosHaveParagraphWithIdBeforeThem() throws Exception {
+        for (Path path : XdocUtil.getXdocsTemplatesFilePaths()) {
+            final String fileName = path.getFileName().toString();
+            final String input = Files.readString(path);
+            final Document document = XmlUtil.getRawXml(fileName, input, input);
+            final NodeList sources = document.getElementsByTagName("macro");
+
+            for (int position = 0; position < sources.getLength(); position++) {
+                final Node macro = sources.item(position);
+                final String macroName = macro.getAttributes()
+                        .getNamedItem("name").getTextContent();
+
+                if (!"example".equals(macroName)) {
+                    continue;
+                }
+
+                Node precedingParagraph = macro.getPreviousSibling();
+                while (!precedingParagraph.getNodeName().equals("p")) {
+                    precedingParagraph = precedingParagraph.getPreviousSibling();
+                }
+
+                assertWithMessage("Paragraph before example macro should have an id attribute")
+                        .that(precedingParagraph.hasAttributes())
+                        .isTrue();
+
+                final Node idAttribute = precedingParagraph.getAttributes().getNamedItem("id");
+                assertWithMessage("Paragraph before example macro should have an id attribute")
+                        .that(idAttribute)
+                        .isNotNull();
+
+                String exampleName = "";
+                String exampleType = "";
+                final NodeList params = macro.getChildNodes();
+                for (int paramPosition = 0; paramPosition < params.getLength(); paramPosition++) {
+                    final Node item = params.item(paramPosition);
+
+                    if (!item.getNodeName().equals("param")) {
+                        continue;
+                    }
+
+                    final String paramName = item.getAttributes()
+                            .getNamedItem("name").getTextContent();
+                    final String paramValue = item.getAttributes()
+                            .getNamedItem("value").getTextContent();
+                    if ("path".equals(paramName)) {
+                        exampleName = paramValue.substring(paramValue.lastIndexOf('/') + 1,
+                                paramValue.lastIndexOf('.'));
+                    }
+                    else if ("type".equals(paramName)) {
+                        exampleType = paramValue;
+                    }
+                }
+
+                final String id = idAttribute.getTextContent();
+                final String expectedId = String.format(Locale.ROOT, "%s-%s", exampleName,
+                        exampleType);
+                assertWithMessage("Paragraph before example macro should have an id attribute")
```

Comment:
When it makes sense, different assertions should have different messages, otherwise, if a user hits one and then another, all they see is the same message and think they didn't resolve the issue when it is really a different issue.

`"Paragraph before example macro should have an id attribute"` => `"Paragraph before example macro should have the expected id value"`

Reply:
Done.